src/List.h

```diff
@@ -20,67 +20,361 @@
 
-	void sort(list_cmp_func cmp_func);
+		if ( new_size != max_entries )
+			{
+			entries = (T*) safe_realloc((void*) entries, sizeof(T) * new_size);
+			if ( entries )
+				max_entries = new_size;
+			else
+				max_entries = 0;
+			}
+
+		return max_entries;
+		}
+
+	void sort(list_cmp_func cmp_func)
+		{
+		qsort(entries, num_entries, sizeof(T), cmp_func);
+		}
 
 	int MemoryAllocation() const
-		{ return padded_sizeof(*this) + pad_size(max_entries * sizeof(ent)); }
+		{ return padded_sizeof(*this) + pad_size(max_entries * sizeof(T)); }
 
-protected:
-	~BaseList()		{ free(entry); }
-	explicit BaseList(int = 0);
-	BaseList(const BaseList&);
-	BaseList(BaseList&&);
-	BaseList(const ent* arr, int n);
+	void insert(const T& a)	// add at head of list
+		{
+		if ( num_entries == max_entries )
+			resize(max_entries ? max_entries * LIST_GROWTH_FACTOR : DEFAULT_LIST_SIZE);
 
-	BaseList& operator=(const BaseList&);
-	BaseList& operator=(BaseList&&);
+		for ( int i = num_entries; i > 0; --i )
+			entries[i] = entries[i-1];	// move all pointers up one
 
-	void insert(ent);	// add at head of list
+		++num_entries;
+		entries[0] = a;
+		}
 
 	// Assumes that the list is sorted and inserts at correct position.
-	void sortedinsert(ent, list_cmp_func cmp_func);
+	void sortedinsert(const T& a, list_cmp_func cmp_func)
+		{
+		// We optimize for the case that the new element is
+		// larger than most of the current entries.
+
+		// First append element.
+		if ( num_entries == max_entries )
+			resize(max_entries ? max_entries * LIST_GROWTH_FACTOR : DEFAULT_LIST_SIZE);
+
+		entries[num_entries++] = a;
+
+		// Then move it to the correct place.
+		T tmp;
+		for ( int i = num_entries - 1; i > 0; --i )
+			{
+			if ( cmp_func(entries[i],entries[i-1]) <= 0 )
+				break;
+
+			tmp = entries[i];
+			entries[i] = entries[i-1];
+			entries[i-1] = tmp;
+			}
+		}
+
+	void push_back(const T& a)	{ append(a); }
+	void push_front(const T& a)	{ insert(a); }
+	T pop_front() { return get() ;}
+	T pop_back() { return remove_nth(num_entries-1); }
+
+	void append(const T& a)	// add to end of list
+		{
+		if ( num_entries == max_entries )
+			resize(max_entries ? max_entries * LIST_GROWTH_FACTOR : DEFAULT_LIST_SIZE);
+
+		entries[num_entries++] = a;
+		}
+
+	T remove(const T& a)	// delete entry from list
+		{
+		int i;
+		for ( i = 0; i < num_entries && a != entries[i]; ++i )
+			;
+
+		return remove_nth(i);
+		}
+
+	T remove_nth(int n)	// delete nth entry from list
+		{
+		if ( n < 0 || n >= num_entries )
+			return 0;
+
+		T old_ent = entries[n];
+		--num_entries;
+
+		for ( ; n < num_entries; ++n )
+			entries[n] = entries[n+1];
 
-	void append(ent);	// add to end of list
-	ent remove(ent);	// delete entry from list
-	ent remove_nth(int);	// delete nth entry from list
-	ent get();		// return and remove ent at end of list
-	ent last()		// return at end of list
-		{ return entry[num_entries-1]; }
+		entries[n] = nullptr;	// for debugging
```

Comment:
Setting it to null existed in the original code: https://github.com/zeek/zeek/blob/master/src/List.cc#L159

It can probably just be removed.

Reply:
Yeah, was thinking just remove -- it was in the original code, but it seems that only worked coincidentally because we only had lists for integer or pointer types?  Assigning a `nullptr` won't work for any arbitrary `T` type.